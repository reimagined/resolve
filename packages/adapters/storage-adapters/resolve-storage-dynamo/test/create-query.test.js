import createQuery from '../src/create-query'
import createAggregateIdExpression from '../src/create-aggregate-id-expression'
import createTypeExpression from '../src/create-type-expression'
import createTimestampExpression from '../src/create-timestamp-expression'
import expressionObject from '../src/expression-object'
import expressionString from '../src/expression-string'

describe('method "createQuery"', () => {
  test('should return correctly query', () => {
    const tableName = 'tableName'

    for (const { aggregateIds, eventTypes, startTime, finishTime } of [
      {
        aggregateIds: undefined,
        eventTypes: undefined,
        startTime: undefined,
        finishTime: undefined
      },
      {
        aggregateIds: null,
        eventTypes: null,
        startTime: undefined,
        finishTime: undefined
      },
      {
        aggregateIds: ['id1'],
        eventTypes: undefined,
        startTime: 5,
        finishTime: undefined
      },
      {
        aggregateIds: ['id1', 'id2'],
        eventTypes: undefined,
        startTime: undefined,
        finishTime: 10
      },
      {
        aggregateIds: ['id1'],
        eventTypes: ['type1'],
        startTime: 5,
        finishTime: 10
      },
      {
        aggregateIds: ['id1', 'id2'],
        eventTypes: ['type1'],
        startTime: undefined,
        finishTime: 10
      },
      {
        aggregateIds: ['id1'],
        eventTypes: ['type1', 'type2'],
        startTime: 5,
        finishTime: undefined
      },
      {
        aggregateIds: ['id1', 'id2'],
        eventTypes: ['type1', 'type2'],
        startTime: undefined,
        finishTime: undefined
      }
    ]) {
      const filter = { aggregateIds, eventTypes, startTime, finishTime }

      const {
        conditionExpression: typeConditionExpression,
        attributeNames: typeAttributeNames,
        attributeValues: typeAttributeValues
      } = createTypeExpression(filter)

      const {
        conditionExpression: timestampConditionExpression,
        attributeNames: timestampAttributeNames,
        attributeValues: timestampAttributeValues
      } = createTimestampExpression(filter)

      const {
        conditionExpression: aggregateIdConditionExpression,
        attributeNames: aggregateIdAttributeNames,
        attributeValues: aggregateIdAttributeValues
      } = createAggregateIdExpression(filter)

      const query = createQuery(
        { tableName, expressionString, expressionObject },
        { aggregateIds },
        {
          aggregateIdAttributeNames,
          aggregateIdAttributeValues,
          aggregateIdConditionExpression,
          typeAttributeNames,
          typeAttributeValues,
          typeConditionExpression,
          timestampAttributeNames,
          timestampAttributeValues,
          timestampConditionExpression
        }
      )

      expect(query).toMatchSnapshot(JSON.stringify(filter))
    }
  })
})
