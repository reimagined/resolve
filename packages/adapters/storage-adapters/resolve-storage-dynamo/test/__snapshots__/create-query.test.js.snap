// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1","id2"],"eventTypes":["type1","type2"]} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#globalPartitionKey": "globalPartitionKey",
    "#type": "type",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":aggregateId1": "id2",
    ":eventType0": "type1",
    ":eventType1": "type2",
    ":globalPartitionKey": "globalPartitionKey",
  },
  "FilterExpression": "(#aggregateId IN (:aggregateId0,:aggregateId1)) AND (#type IN (:eventType0,:eventType1))",
  "IndexName": "rangedEvents",
  "KeyConditionExpression": "(#globalPartitionKey = :globalPartitionKey)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1","id2"],"eventTypes":["type1"],"finishTime":10} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#globalPartitionKey": "globalPartitionKey",
    "#timestamp": "timestamp",
    "#type": "type",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":aggregateId1": "id2",
    ":eventType0": "type1",
    ":finishTime": 10,
    ":globalPartitionKey": "globalPartitionKey",
  },
  "FilterExpression": "(#aggregateId IN (:aggregateId0,:aggregateId1)) AND (#type = :eventType0)",
  "IndexName": "rangedEvents",
  "KeyConditionExpression": "(#globalPartitionKey = :globalPartitionKey) AND (#timestamp < :finishTime)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1","id2"],"finishTime":10} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#globalPartitionKey": "globalPartitionKey",
    "#timestamp": "timestamp",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":aggregateId1": "id2",
    ":finishTime": 10,
    ":globalPartitionKey": "globalPartitionKey",
  },
  "FilterExpression": "(#aggregateId IN (:aggregateId0,:aggregateId1))",
  "IndexName": "rangedEvents",
  "KeyConditionExpression": "(#globalPartitionKey = :globalPartitionKey) AND (#timestamp < :finishTime)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1"],"eventTypes":["type1","type2"],"startTime":5} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#timestamp": "timestamp",
    "#type": "type",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":eventType0": "type1",
    ":eventType1": "type2",
    ":startTime": 5,
  },
  "FilterExpression": "(#type IN (:eventType0,:eventType1)) AND (#timestamp > :startTime)",
  "KeyConditionExpression": "(#aggregateId = :aggregateId0)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1"],"eventTypes":["type1"],"startTime":5,"finishTime":10} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#timestamp": "timestamp",
    "#type": "type",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":eventType0": "type1",
    ":finishTime": 10,
    ":startTime": 5,
  },
  "FilterExpression": "(#type = :eventType0) AND (#timestamp BETWEEN :startTime AND :finishTime)",
  "KeyConditionExpression": "(#aggregateId = :aggregateId0)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {"aggregateIds":["id1"],"startTime":5} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#aggregateId": "aggregateId",
    "#timestamp": "timestamp",
  },
  "ExpressionAttributeValues": Object {
    ":aggregateId0": "id1",
    ":startTime": 5,
  },
  "FilterExpression": "(#timestamp > :startTime)",
  "KeyConditionExpression": "(#aggregateId = :aggregateId0)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;

exports[`method "createQuery" should return correctly query: {} 1`] = `
Object {
  "ExpressionAttributeNames": Object {
    "#globalPartitionKey": "globalPartitionKey",
  },
  "ExpressionAttributeValues": Object {
    ":globalPartitionKey": "globalPartitionKey",
  },
  "FilterExpression": undefined,
  "IndexName": "rangedEvents",
  "KeyConditionExpression": "(#globalPartitionKey = :globalPartitionKey)",
  "ScanIndexForward": true,
  "TableName": "tableName",
}
`;
